Custom s3 endpoint support

pitscher opened the pull request:
As mentioned in #27 this is the PR for supporting custom s3 endpoints.
I successfully ran the e2e script for AWS (because noting GCP related was changed).

If you don't set the new env var AWSCLI_ENDPOINT_URL_OPTION the default s3 endpoint will be used.
If you set the var, the provided endpoint will be used. That's it. No rocket science. 😄 

BTW: I updated the README as well.

pitscher committed:
Added new env var for custom endpoint

pitscher committed:
Removed otc endpoint url and added placeholder

pitscher committed:
Include custom s3 endpoint option in README

ryu-sato reviewed (commented):
Thank you very much for your contributions.

Could you use the word "S3" instead of "s3" in README.md?

functions.sh LGTM.

skomma commented on bin/functions.sh:
```diff
@@ -19,7 +20,7 @@ DATE_CMD="/bin/date"
 s3_exists() {
 	if [ $# -ne 1 ]; then return 255; fi
-	${AWSCLI} ${AWSCLIOPT} ${AWSCLI_LIST_OPT} $1 >/dev/null
+	${AWSCLI} ${AWSCLI_ENDPOINT_URL_OPTION} ${AWSCLIOPT} ${AWSCLI_LIST_OPT} $1 >/dev/null
```
`AWSCLI_ENDPOINT_URL_OPTION` will be set like `fra1.digitaloceanspaces.com`, so `--endpoint-url` is needed here, right?

pitscher replied:
Yes. It's important to set the whole aws cli parameter "--endpoint-url https://domain.tld".
It is easier to implement this way because if a user wanna use the default S3 endpoint the only thing he must ensure is to not set the AWSCLI_ENDPOINT_OPT var. That's it. 😃

skomma commented on bin/functions.sh:
```diff
@@ -1,5 +1,6 @@
+#AWSCLI_ENDPOINT_URL_OPTION="--endpoint-url <YourEndpointUrl>"
 AWSCLI_COPY_OPT="s3 cp"
 AWSCLI_LIST_OPT="s3 ls"
 AWSCLI_DEL_OPT="s3 rm"
```
Since `AWSCLIOPT` is already provided, I think you can do the same thing by using this environment variable without adding `AWSCLI_ENDPOINT_URL_OPTION`.

Unfortunately, currently `AWSCLIOPT` is not documented...

pitscher replied:
To provide the ability to use the default endpoint or a custom one and not over engineer the code we should use a new env var. Therefore I introduced AWSCLI_ENDPOINT_OPT

pitscher committed:
Rename ENDPOINT env var

pitscher committed:
Update README with new renamed ENDPOINT var

pitscher commented:
As mentioned here: https://github.com/weseek/mongodb-awesome-backup/pull/28#discussion_r431868344
I updated a few things like the name of the new ENDPOINT env var and renamed "s3" to "S3".
Now all should be fine. 🤓

pitscher commented:
Hi guys,
could you give me an update regarding my last changes? 😃

ryu-sato reviewed (changes requested):
Comparing the usage and implementation of AWSCLI_ENDPOINT_OPT, README.md lists "fra1.digitaloceanspaces.com" in the input example, but functions.sh expects user to specify "--endpoint-url" as in "--endpoint-url fra1.digitaloceanspaces.com", so you need to match two.

How to modify it...

1. Change the input example of AWSCLI_ENDPOINT_OPT to "--endpoint-url fra1.digitaloceanspaces.com".
2. In functions.sh, if AWSCLI_ENDPOINT_OPT exists, pass "--endpoint-url $AWSCLI_ENDPOINT_OPT" to `aws` options.
3. Delete AWSCLI_ENDPOINT_OPT from README.md and functions.sh, and describe "--endpoint-url fra1.digitaloceanspaces.com" as an example of AWSCLIOPT input.

1 is unacceptable because the name AWSCLI_ENDPOINT_OPT is different from the actual behavior because it allows user to specify an AWS option other than "--endpoint-url".
2 is closer to the current implementation than 3, so can you modify it with this policy?

Also, please modify the following points in the README (If you chose policy 2)

* Make the input example of AWSCLI_ENDPOINT_OPT into a URL
    * ex. "AWSCLI_ENDPOINT_OPT="https://fra1.digitaloceanspaces.com"
* Add AWSCLI_ENDPOINT_OPT to the three docker run command execution examples in Usage
    * ex. `[ -e AWSCLI_ENDPOINT_OPT=<S3 endpoint URL (ex. https://fra1.digitaloceanspaces.com)> \ ]`

pitscher committed:
Remove old comment

pitscher committed:
Add check if custom aws endpoint is set

pitscher committed:
Add custom endpoint option to docker run examples

pitscher commented:
Thank you for the feedback 😃 
I've implemented what you requested. Hopefully this correspond to your 2. policy.

ryu-sato commented on bin/functions.sh:
```diff
@@ -19,7 +19,12 @@ DATE_CMD="/bin/date"
+			${AWSCLI} ${AWSCLIOPT} ${AWSCLI_LIST_OPT} $1 >/dev/null
+		else
+			${AWSCLI} --endpoint-url ${AWSCLI_ENDPOINT_OPT} ${AWSCLIOPT} ${AWSCLI_LIST_OPT} $1 >/dev/null
+		fi
```
```suggestion
	${AWSCLI} ${AWSCLI_ENDPOINT_OPT} ${AWSCLIOPT} ${AWSCLI_LIST_OPT} $1 >/dev/null
```

ryu-sato commented on bin/functions.sh:
```diff
@@ -29,7 +34,12 @@ gs_exists() {
+		${AWSCLI} ${AWSCLIOPT} ${AWSCLI_LIST_OPT} $1
+	else
+		${AWSCLI} --endpoint-url ${AWSCLI_ENDPOINT_OPT} ${AWSCLIOPT} ${AWSCLI_LIST_OPT} $1
+	fi
```
```suggestion
  ${AWSCLI} ${AWSCLI_ENDPOINT_OPT} ${AWSCLIOPT} ${AWSCLI_LIST_OPT} $1
```

ryu-sato commented on bin/functions.sh:
```diff
@@ -39,7 +49,12 @@ gs_list_files() {
+			${AWSCLI} ${AWSCLIOPT} ${AWSCLI_DEL_OPT} $1
+		else
+			${AWSCLI} --endpoint-url ${AWSCLI_ENDPOINT_OPT} ${AWSCLIOPT} ${AWSCLI_DEL_OPT} $1
+		fi
```
```suggestion
  ${AWSCLI} ${AWSCLI_ENDPOINT_OPT} ${AWSCLIOPT} ${AWSCLI_DEL_OPT} $1
```

ryu-sato commented on bin/functions.sh:
```diff
@@ -57,7 +72,11 @@ gs_delete_file() {
+		then
+			${AWSCLI} ${AWSCLIOPT} ${AWSCLI_COPY_OPT} $1 $2
+		else
+			${AWSCLI} --endpoint-url ${AWSCLI_ENDPOINT_OPT} ${AWSCLIOPT} ${AWSCLI_COPY_OPT} $1 $2
```
There is no "fi" on line 80 and it is a syntax error.

```suggestion
  ${AWSCLI} ${AWSCLI_ENDPOINT_OPT} ${AWSCLIOPT} ${AWSCLI_COPY_OPT} $1 $2
```

ryu-sato reviewed (changes requested):
We want the code to be DRY.

It is better not to use `if` many times according to the value of AWSCLI_ENDPOINT_OPT.
If you set it to a variable beforehand, you don't have to repeat using `if`.

In addition to the suggestions I commented on, please add the following to line 6 of "bin/functions.sh"

```bash
AWSCLI_ENDPOINT_OPT=${AWSCLI_ENDPOINT_OPT:+"--endpoint-url ${AWSCLI_ENDPOINT_OPT}"}
```

pitscher committed:
Remove repetitive if-statements

pitscher commented:
Of course. Good point! This should finally be what you expected.
Thank you for this sleek tip. I mean... I created the PR and you invested a bunch of your time so far.
But I appreciate your help. I learned a lot during this PR. 😄

ryu-sato reviewed (approved):
LGTM
